src/uct/cuda/base/cuda_md.c

```diff
@@ -155,6 +155,13 @@ uct_cuda_base_query_attributes(uct_cuda_copy_md_t *md, const void *address,
         goto out_default_range;
     }
 
+    cu_err = cudaSetDevice(cuda_device);
+    if (cu_err != CUDA_SUCCESS) {
+        cuGetErrorString(cu_err, &cu_err_str);
+        ucs_error("cudaSetDevice(%d) error: %s", cuda_device, cu_err_str);
+        return UCS_ERR_IO_ERROR;
+    }
+
```

Comment:
how is this related?

Reply:
goroutines failing without this

Reply:
in general we should not do this, to not affect user applications
what is the failure?

Reply:
@Akshay-Venkatesh how could it be that cuPointerGetAttributes detected device memory, but cuMemGetAddressRange failed later?

Reply:
Apologies I've missed the mention @yosefe @gleon99 . Are the the threads calling cuPointerGetAttributes and cuMemGetAddressRange one and the same? Calling cuPointerGetAttributes doesn't require a thread to have an active context set but I believe calling cuMemGetAddressRange did require device context to associated with the thread. As the error suggests maybe there is no context associated with the thread?

Reply:
@Artemy-Mellanox need to resolve this as well - we should not cudaSetDevice in UCX. Maybe just in Go test.

Reply:
What if user will get same error we get in Go test? That's what I found:
> Make it a habit to call cudaSetDevice() wherever your code could potentially spawn new host threads.

https://developer.nvidia.com/blog/cuda-pro-tip-always-set-current-device-avoid-multithreading-bugs/

Maybe we should do it only when UCS_THREAD_MODE_MULTI?

Reply:
1. can we set cuda device from Golang tests?
2. what has changed that is causing this - do we call uct_cuda_base_query_attributes() earlier than before this PR?
@petro-rudenko WDYT?

Reply:
@yosefe - golang test doesn't use cuda API directly. It only uses `ucp_mem_map` with memtype CUDA. Setting cudaDevice from tests would require additional dependency for go - would also need to whether use some cuda API for go, or to make additional wrappers, etc.

Reply:
@petro-rudenko do you know how the device is set then? AFAIR needed to call some UCX API to implicitly set the device?

Reply:
1. use driver API
2. set device only if address range query fails with "no device"

Reply:
We shouldn't be setting cuda device in uct/cuda at any point. If the user has an intent to allocate memory on a specific device, the device must be selected in application land. IMO, if at all we want to do this in uct, we should expose an attribute in uct memmap API and set/unset per invocation.

Reply:
@Akshay-Venkatesh the problem is if the memory was allocated by thread A, which does have a context set, and then used in UCX API by thread B - that calls cu query range and fails

Reply:
For this to work, either thread B has to set the correct context (as we force applications to do today) or we need cache cuda context with a memory range and allow any thread to query a subrange by setting/unsetting context before/after query. Are there other less invasive options?

Reply:
we could know which gpu is the owner of the relevant VA by cuda pointer query, and then set the context only if it wasn't set already for this thread.
If the user explicitly sets a different cuda context - it will not be supported.

Reply:
sounds good to me